### PR TITLE
Feat/router hop

### PR DESCRIPTION
**Pull request by alaibe**
* Add bridge hop integration
** Add contract
** Calculate fees
* Update router algorithm
** Find best route
** Find possible candidate
** Add token USDC for l2 chains

### CODE REVIEW

File: services/wallet/bridge/hop.go [hunk 1]
```diff
@@ -0,0 +1,378 @@
...
+	"github.com/status-im/status-go/services/wallet/token"
+	"github.com/status-im/status-go/transactions"
+)
+
+const HopLpFeeBps = 4
+const HopCanonicalTokenIndex = 0
+const HophTokenIndex = 1
+const HopMinBonderFeeUsd = 0.25
+
+var HopBondTransferGasLimit = map[uint64]int64{
```
Comment (Khushboo-dev-cpp) [inline services/wallet/bridge/hop.go:27]: these are hardcoded? where do we get these values from?

File: services/wallet/bridge/hop.go [hunk 2]
```diff
@@ -0,0 +1,378 @@
...
+		return false, nil
+	}
+
+	if _, ok := fees[to.ChainID]; !ok {
+		return false, nil
+	}
+	return true, nil
+}
+
+func (s *HopBridge) EstimateGas(from, to *params.Network, token *token.Token, amountIn *big.Int) (uint64, error) {
```
Comment (Khushboo-dev-cpp) [inline services/wallet/bridge/hop.go]: this is wip as far as I know

File: services/wallet/bridge/hop.go [hunk 3]
```diff
@@ -0,0 +1,378 @@
...
+	}
+	amountOutInt, _ := amountOut.Int(nil)
+
+	return bonderFees, new(big.Int).Add(
+		bonderFees,
+		new(big.Int).Sub(amountIn, amountOutInt),
+	), nil
+}
+
+func (h *HopBridge) calcToHTokenAmount(network *params.Network, amount *big.Float, symbol string) (*big.Float, error) {
```
Comment (Khushboo-dev-cpp) [inline services/wallet/bridge/hop.go:293]: this name of function is not understandable to me

File: services/wallet/bridge/hop.go [hunk 4]
```diff
@@ -0,0 +1,378 @@
...
+	hTokenAmount, err := h.calcToHTokenAmount(from, amount, token.Symbol)
+	if err != nil {
+		return nil, err
+	}
+	feeBps := HopBonderFeeBps[token.Symbol][to.ChainID]
+
+	factor := new(big.Float).Mul(hTokenAmount, big.NewFloat(float64(feeBps)))
+	return new(big.Float).Quo(
+		factor,
+		big.NewFloat(10000),
```
Comment (Khushboo-dev-cpp) [inline services/wallet/bridge/hop.go:359]: can this value be a constant?

File: services/wallet/bridge/simple.go [hunk 5]
```diff
@@ -0,0 +1,47 @@
...
+	return from.ChainID == to.ChainID, nil
+}
+
+func (s *SimpleBridge) CalculateFees(from, to *params.Network, token *token.Token, amountIn *big.Int, nativeTokenPrice, tokenPrice float64, gasPrice *big.Float) (*big.Int, *big.Int, error) {
+	return big.NewInt(0), big.NewInt(0), nil
+}
+
+func (s *SimpleBridge) EstimateGas(from, to *params.Network, token *token.Token, amountIn *big.Int) (uint64, error) {
+	if token.IsNative() {
+		return 22000, nil // default gas limit for eth transaction
```
Comment (Khushboo-dev-cpp) [inline services/wallet/bridge/simple.go:36]: these as well?

File: services/wallet/router.go [hunk 6]
```diff
@@ -7,49 +7,197 @@ import (
...
 
-	return false, nil
+	if n.Path == nil {
+		return best, bestTotalCost
+	}
+
+	return append([]*Path{n.Path}, best...), new(big.Float).Add(bestTotalCost, n.Path.Cost)
+}
+
+type SuggestedRoutes struct {
```
Comment (Khushboo-dev-cpp) [inline services/wallet/router.go:188]: can this move on top with the other type definitions?

**Review by Khushboo-dev-cpp — APPROVED** [review]
Such good work on this :)

just some very minor comments!

File: services/wallet/bridge/hop.go (hunk 3 above)
Comment (alaibe) [reply services/wallet/bridge/hop.go:293]: This is the algorithm provided by hop, I will provide a link to it

File: services/wallet/bridge/hop.go (hunk 2 above)
Comment (alaibe) [reply services/wallet/bridge/hop.go]: I couldn't get estimate to work properly, I will add a todo

File: services/wallet/bridge/hop.go (hunk 1 above)
Comment (alaibe) [reply services/wallet/bridge/hop.go:27]: From hop config, I ll add a link to find out where they come from

File: services/wallet/bridge/hop.go (hunk 3 above)
Comment (Khushboo-dev-cpp) [reply services/wallet/bridge/hop.go:293]: hey no I meant the name calcToHTokenAmount? do we need to keep it this way?

File: services/ens/api.go [hunk 7]
```diff
@@ -337,12 +337,16 @@ func (api *API) ReleaseEstimate(ctx context.Context, chainID uint64, txArgs tran
...
+	estimate, err := ethClient.EstimateGas(ctx, ethereum.CallMsg{
 		From:  common.Address(txArgs.From),
 		To:    &registryAddr,
 		Value: big.NewInt(0),
 		Data:  data,
 	})
+	if err != nil {
+		return 0, err
+	}
+	return estimate + 1000, nil
```
Comment (richard-ramos) [inline services/ens/api.go:358]: :100:

File: services/wallet/router.go [hunk 8]
```diff
@@ -2,54 +2,276 @@ package wallet
...
 }
 
-type Router struct {
-	s *Service
+func (s SendType) isAvailableFor(network *params.Network) bool {
+	if s == Transfer {
+		return true
+	}
+
+	if network.ChainID == 1 || network.ChainID == 5 {
```
Comment (richard-ramos) [inline services/wallet/router.go:43]: Can these chainID be replaced by constants?

File: services/wallet/router.go [hunk 9]
```diff
@@ -2,54 +2,276 @@ package wallet
...
+
+	return false
 }
 
-func (r *Router) suitableTokenExists(ctx context.Context, network *params.Network, tokens []*Token, account common.Address, amount float64, tokenSymbol string) (bool, error) {
-	for _, token := range tokens {
-		if token.Symbol != tokenSymbol {
-			continue
+func (s SendType) EstimateGas(service *Service, network *params.Network) uint64 {
+	from := types.Address(common.HexToAddress("0x5ffa75ce51c3a7ebe23bde37b5e3a0143dfbcee0"))
```
Comment (richard-ramos) [inline services/wallet/router.go:51]: What is this contract? Could it be replaced by a constant as well? Is it available in all chains?

File: services/wallet/router.go [hunk 10]
```diff
@@ -2,54 +2,276 @@ package wallet
...
+	tx := transactions.SendTxArgs{
+		From:  from,
+		Value: (*hexutil.Big)(big.NewInt(0)),
+	}
+	if s == ENSRegister {
+		estimate, err := service.ens.API().RegisterEstimate(context.Background(), network.ChainID, tx, EstimateUsername, EstimatePubKey)
+		if err != nil {
+			return 400000
```
Comment (richard-ramos) [inline services/wallet/router.go:59]: I wonder if we should return the error too. EstimateGas generally fails if the transaction will revert, altho we should still give the user the possibility of sending the transaction

File: services/wallet/router.go [hunk 11]
```diff
@@ -2,54 +2,276 @@ package wallet
...
 		}
+		return estimate
+	}
 
-		clients, err := chain.NewClients(r.s.rpcClient, []uint64{network.ChainID})
+	if s == ENSRelease {
+		estimate, err := service.ens.API().ReleaseEstimate(context.Background(), network.ChainID, tx, EstimateUsername)
 		if err != nil {
-			return false, err
+			return 200000
```
Comment (richard-ramos) [inline services/wallet/router.go:67]: Same as previous comment

File: services/wallet/router.go [hunk 12]
```diff
@@ -2,54 +2,276 @@ package wallet
...
 		}
+		return estimate
+	}
 
-		balance, err := r.s.tokenManager.getBalance(ctx, clients[0], account, token.Address)
+	if s == ENSSetPubKey {
+		estimate, err := service.ens.API().SetPubKeyEstimate(context.Background(), network.ChainID, tx, fmt.Sprint(EstimateUsername, ".stateofus.eth"), EstimatePubKey)
 		if err != nil {
-			return false, err
+			return 400000
```
Comment (richard-ramos) [inline services/wallet/router.go:75]: =

File: services/wallet/router.go [hunk 13]
```diff
@@ -2,54 +2,276 @@ package wallet
...
+		}
+		return estimate
+	}
+
+	if s == StickersBuy {
+		packId := &bigint.BigInt{Int: big.NewInt(2)}
+		estimate, err := service.stickers.API().BuyEstimate(context.Background(), network.ChainID, from, packId)
+		if err != nil {
+			return 400000
```
Comment (richard-ramos) [inline services/wallet/router.go:84]: =

File: services/wallet/router.go [hunk 14]
```diff
@@ -2,54 +2,276 @@ package wallet
...
+type Node struct {
+	Path     *Path
+	Children Graph
+}
+
+func newNode(path *Path) *Node {
+	return &Node{Path: path, Children: make(Graph, 0)}
+}
+
+func buildGraph(AmountIn *big.Int, routes []*Path, level int, sourceChainIDs []uint64) Graph {
```
Comment (richard-ramos) [inline services/wallet/router.go:121]: :1st_place_medal:

File: services/wallet/bridge/simple.go [hunk 15]
```diff
@@ -0,0 +1,47 @@
...
+	}
+
+	return 200000, nil //default gas limit for erc20 transaction
```
Comment (richard-ramos) [inline services/wallet/bridge/simple.go:39]: Shouldnt these values be estimated instead of hardcoded?

**Review by richard-ramos — APPROVED** [review]

File: services/wallet/bridge/simple.go (hunk 15 above)
Comment (alaibe) [reply services/wallet/bridge/simple.go:39]: We do not know the contract for each erc20, so it means it is not stable, I put a top value I could find

**Pull request merged**